sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -150,138 +118,232 @@ private[sql] object OrcFilters extends OrcFiltersBase {
+      builder: Builder
+  ): Option[Builder] = {
     def getType(attribute: String): PredicateLeaf.Type =
       getPredicateLeafType(dataTypeMap(attribute))
 
     import org.apache.spark.sql.sources._
 
-    expression match {
-      case And(left, right) =>
-        // At here, it is not safe to just convert one side and remove the other side
-        // if we do not understand what the parent filters are.
-        //
-        // Here is an example used to explain the reason.
-        // Let's say we have NOT(a = 2 AND b in ('1')) and we do not understand how to
-        // convert b in ('1'). If we only convert a = 2, we will end up with a filter
-        // NOT(a = 2), which will generate wrong results.
-        //
-        // Pushing one side of AND down is only safe to do at the top level or in the child
-        // AND before hitting NOT or OR conditions, and in this case, the unsupported predicate
-        // can be safely removed.
-        val leftBuilderOption =
-          createBuilder(dataTypeMap, left, newBuilder, canPartialPushDownConjuncts)
-        val rightBuilderOption =
-          createBuilder(dataTypeMap, right, newBuilder, canPartialPushDownConjuncts)
-        (leftBuilderOption, rightBuilderOption) match {
-          case (Some(_), Some(_)) =>
-            for {
-              lhs <- createBuilder(dataTypeMap, left,
-                builder.startAnd(), canPartialPushDownConjuncts)
-              rhs <- createBuilder(dataTypeMap, right, lhs, canPartialPushDownConjuncts)
-            } yield rhs.end()
-
-          case (Some(_), None) if canPartialPushDownConjuncts =>
-            createBuilder(dataTypeMap, left, builder, canPartialPushDownConjuncts)
+    // The performAction method can run both the filtering and building operations for a given
+    // node - we signify which one we want with the `actionType` parameter.
+    //
+    // There are a couple of benefits to coupling the two operations like this:
+    // 1. All the logic for a given predicate is grouped logically in the same place. You don't
+    //   have to scroll across the whole file to see what the filter action for an And is while
+    //   you're looking at the build action.
+    // 2. It's much easier to keep the implementations of the two operations up-to-date with
+    //   each other. If the `filter` and `build` operations are implemented as separate case-matches
+    //   in different methods, it's very easy to change one without appropriately updating the
+    //   other. For example, if we add a new supported node type to `filter`, it would be very
+    //   easy to forget to update `build` to support it too, thus leading to conversion errors.
+    //
+    // Doing things this way does have some annoying side effects:
+    // - We need to return an `Either`, with one action type always returning a Left and the other
+    //   always returning a Right.
+    // - We always need to pass the canPartialPushDownConjuncts parameter even though the build
```

Comment:
one idea is to carry the `canPartialPushDownConjuncts` in its corresponding action.

Reply:
Yep, that's a good idea - implemented.

I had a similar thought around type members per action, but couldn't get it to work so kind of gave up on putting more things there.